Joda formats for Scanamo

philwills opened the pull request:
Fun: the tests run fine on Travis, but locally jodaStringFormat blows up immediately though testing it at the console seems fine. 

Exactly the kind of 'fun' you expect with dates.

If anyone has the chance to try it locally and see if it works for them, I'd appreciate knowing if it's just a  quirk of my machine.

Partially addresses #203

regiskuckaertz commented:
Same here. There seems to be two problems.

Just ran locally and it didn't pass in this case:
```
[info] ! JodaFormats.scala.jodaStringFormat.property at line 13: (dt: DateTime) => ...: Falsified after 0 passed tests.
[info] > ARG_0: -57277773-09-24T23:24:30.896-00:01:15
```
Interestingly:
```
scala> jodaStringFormat.write(res0)
res2: com.amazonaws.services.dynamodbv2.model.AttributeValue = {S: -57277773-09-24T23:24:30.896-00:01:15,}

scala> jodaStringFormat.read(res2)
res3: Either[com.gu.scanamo.error.DynamoReadError,org.joda.time.DateTime] = Right(-57277773-09-24T23:25:45.896Z)
```
The problem is `.withZone` normalises the above example to the specific timezone. I suggest removing it, as it is the client's responsibility to decide which time zone is sensible for their use case. (I tested without and it worked fine)

However, even with that fix, the test is still failing for me. My assumption is that the `toString` function does not produce a faithful rendering of the value, it may be missing some information. From the [source code of `toString`](https://github.com/JodaOrg/joda-time/blob/master/src/main/java/org/joda/time/base/AbstractDateTime.java#L313):
> Output the date time in ISO8601 format (yyyy-MM-ddTHH:mm:ss.SSSZZ).
> Note that this method does not output the chronology or time-zone. This can be confusing, as the equals and hashCode methods use both chronology and time-zone. If two objects are not {@code equal} but have the same {@code toString} then either the chronology or time-zone differs.

I would suggest using the `toString(pattern)` flavour instead.

philwills commented:
I've removed the normalisation in the format and got the test to pass in a consistent Chronology and TimeZone and added a note to the Scala doc that ensuring that is left as an exercise for the reader.

philwills commented on joda/src/main/scala/com/gu/scanamo/joda/JodaFormats.scala:
```diff
@@ -0,0 +1,43 @@
+    *      | DynamoFormat[DateTime].read(DynamoFormat[DateTime].write(dt)) == Right(dt)
+    *  }}}
+    */
+  implicit val jodaEpochSecondsFormat = DynamoFormat.coercedXmap[DateTime, Long, IllegalArgumentException](
```
I'm of the school that locally scoped type classes aren't evil and ensuring you import the right one is a reasonable requirement.

I think this is one of those cases where trying to follow the haskell precedent doesn't bring enough benefit for the syntactical overhead.
